### PR TITLE
Batch processing of revisions in APIExtractor (and generalizations that made it possible)

### PR DESCRIPTION
**Pull request by halfak**
This pull request includes a set of changes to enable batch extraction of features from a set of revision.  This includes a slight refactoring of the dependent.py to allow for a 'context' to be set.  This allows the injection of new dependency processors in place of old ones.  APIExtractor and Language now tie into this new system.  These changes will pave the way for a hybrid DBAPIExtractor.  

You'll also notice a minor changes across the board as I was making repairs to tests and ensure test coverage post-refactoring.

### CODE REVIEW

File: revscoring/datasources/tests/test_page_creation.py [hunk 1]
```diff
@@ -1,28 +1 @@
-from deltas import Delete, Equal, Insert
-from mw import Timestamp
-from nose.tools import eq_
-
 from .. import page_creation
```
Comment (he7d3r) [inline revscoring/datasources/tests/test_page_creation.py]: Is this still necessary?

File: revscoring/datasources/tests/test_previous_user_revision.py [hunk 2]
```diff
@@ -2,30 +2,3 @@
 from nose.tools import eq_
 
 from .. import previous_user_revision
```
Comment (he7d3r) [inline revscoring/datasources/tests/test_previous_user_revision.py]: Is this file still used for anything?

File: revscoring/datasources/tests/test_site.py [hunk 3]
```diff
@@ -1,45 +1,3 @@
 from nose.tools import eq_
 
 from .. import site
```
Comment (he7d3r) [inline revscoring/datasources/tests/test_site.py]: Is this file still used for anything now that it has only import commands?

File: revscoring/datasources/tests/test_types.py [hunk 4]
```diff
@@ -2,91 +2,3 @@
 from nose.tools import eq_
 
 from ..types import RevisionMetadata, UserInfo
```
Comment (he7d3r) [inline revscoring/datasources/tests/test_types.py]: Is this file still used for anything now that it has only import commands?

File: revscoring/datasources/user.py [hunk 5]
```diff
@@ -2,27 +2,4 @@
 from .datasource import Datasource
 from .types import UserInfo
```
Comment (he7d3r) [inline revscoring/datasources/user.py]: Looks like `revision` and `UserInfo` are not used anymore.

File: revscoring/languages/tests/test_persian.py [hunk 6]
```diff
@@ -1,6 +1,6 @@
 from nose.tools import eq_
 
-from ..farsi import is_misspelled
+from ..persian import is_misspelled
```
Comment (he7d3r) [inline revscoring/languages/tests/test_persian.py]: Maybe you should update the "Notafarsiword" below to "Notapersianword" for consistency?

File: revscoring/scorers/scorer.py [hunk 7]
```diff
@@ -24,15 +25,28 @@ def __init__(self, model_map, extractor):
...
-        # If a set of models isn't specified, score with all of 'em
+    def dependencies(self, models=None):
+        return expand_many(self.features(models))
+
+    def features(self, models=None):
+        """
+        Gathers a single tuple of unique features needed by the models
+        """
+        # If no particular model is requested, generate a for all available
+        # models
```
Comment (he7d3r) [inline revscoring/scorers/scorer.py]: "generate a"?

File: revscoring/scorers/scorer.py [hunk 8]
```diff
@@ -24,15 +25,28 @@ def __init__(self, model_map, extractor):
...
+        models = models or self.model_map.keys()
+
+        return tuple({feature for name in models
+                      for feature in self.model_map[name].features})
+
+    def score(self, rev_id, models=None, cache=None):
+        # If no particular model is requested, generate a for all available
```
Comment (he7d3r) [inline revscoring/scorers/scorer.py]: "generate a"?

File: revscoring/tests/test_dependent.py [hunk 9]
```diff
@@ -1,20 +1,83 @@
...
+    my_foo = Dependent("foo", lambda: "foo")
+    foobar = Dependent("foobar", lambda foo: foo + "bar", depends_on=[foo])
+
+    # A function with no arguements
+    eq_(solve(foobar, context={foo: lambda: "foo"}), "foobar")
+
+    # A set of Dependent
+    eq_(solve(foobar, context={my_foo}), "foobar")
+
+    # A dist of Dependent
```
Comment (he7d3r) [inline revscoring/tests/test_dependent.py]: dist -> dict?

File: revscoring/extractors/api.py [hunk 10]
```diff
@@ -13,19 +27,300 @@ def __init__(self, session, language=None):
...
+            rev_id = revision_metadata.parent_id
+            try:
+                logger.info("Requesting a parent revision ({0}) from the API" \
+                            .format(rev_id))
+                return self.session.revisions.get(
+                    rev_id=rev_id,
+                    properties=props
+                )
+            except KeyError:
+                return None
```
Comment (he7d3r) [inline revscoring/extractors/api.py]: Shouldn't this be a `RevisionDocumentNotFound` as in `process_revision_doc`?

File: revscoring/extractors/api.py [hunk 11]
```diff
@@ -13,19 +27,300 @@ def __init__(self, session, language=None):
...
+            docs = list(docs)
+
+            if len(docs) > 0:
+                return docs[0]
+            else:
+                return None
+        else:
+            return None
+
+    def process_page_creation_doc(self, revision_metadata):
```
Comment (he7d3r) [inline revscoring/extractors/api.py]: Duplicated name (see line 177 above)

File: revscoring/extractors/api.py [hunk 12]
```diff
@@ -13,19 +27,300 @@ def __init__(self, session, language=None):
...
+                                rev_doc.get('size'),
+                                'minor' in rev_doc)
+
+    @classmethod
+    def user_info_from_doc(cls, user_doc):
+        if user_doc is None: return None
+        try:
+            registration = Timestamp(user_doc.get('registration'))
+        except ValueError:
+            registration = None
```
Comment (he7d3r) [inline revscoring/extractors/api.py]: Is there any reason for using try/except here, but using if/else on line 267-270? Is any of the two methods preferred over the other?

**Comment by he7d3r** [comment]
@halfak: I'm {{done}} reviewing this version.

File: revscoring/datasources/tests/test_page_creation.py (hunk 1 above)
Comment (halfak) [inline revscoring/datasources/tests/test_page_creation.py]: My thought is that, even if there is nothing to test within the module, this test will still make sure we can import it.  Same to the others below.

File: revscoring/languages/tests/test_persian.py (hunk 6 above)
Comment (halfak) [inline revscoring/languages/tests/test_persian.py]: I don't know.  @Ladsgroup, what should the test case say when demonstrating a misspelling?

File: revscoring/extractors/api.py (hunk 10 above)
Comment (halfak) [inline revscoring/extractors/api.py]: It's common to not have a parent revision.  Datasources/features are expected to handle the lack of a parent_revision.  This is how we handled parent revisions before.

File: revscoring/extractors/api.py (hunk 12 above)
Comment (halfak) [inline revscoring/extractors/api.py]: Should be try/catch in both places.  Fixed above too.

**Comment by halfak** [comment]
@he7d3r updates pushed.  Can you take another look?

File: revscoring/extractors/api.py (hunk 11 above)
Comment (halfak) [inline revscoring/extractors/api.py]: Removed duplicate above.

**Comment by he7d3r** [comment]
I was getting
> ERROR: Failure: DictNotFoundError (Dictionary for language 'fr' could not be found)

but
> $ sudo apt-get install myspell-fr

solved the problem (this should be added to the installation notes). Now the tests pass (except for https://github.com/halfak/Revision-Scoring/issues/65, which also fails on master).

**Comment by halfak** [comment]
Done!

**Comment by halfak** [comment]
@he7d3r

**Comment by halfak** [comment]
Maybe if @he7d3r is not available @ToAruShiroiNeko can merge since he7d3r has signed off and offered to merge before I fixed the fr issue in the readme.

**Pull request merged**